prov/util/src/util_coll.c

```diff
@@ -76,11 +76,20 @@ int ofi_av_set_union(struct fid_av_set *dst, const struct fid_av_set *src)
 				break;
 		}
 		if (j == dst_av_set->fi_addr_count) {
-			dst_av_set->fi_addr_array[temp_count++] =
-				src_av_set->fi_addr_array[i];
+			if ((temp_count+1) >= dst_av_set->max_array_size) {
+				dst_av_set->fi_addr_count += src_av_set->fi_addr_count;
+				dst_av_set->fi_addr_array = (fi_addr_t *)realloc(
+							dst_av_set->fi_addr_array,
+							dst_av_set->fi_addr_count * sizeof(fi_addr_t));
+				if (!dst_av_set->fi_addr_array)
```

Comment:
The way to handle realloc calls is to save the results to another variable.  That way if the call fails, you don't lose the existing array completely.